src/mesh.cpp

```diff
@@ -1373,6 +1403,443 @@ bool RectilinearMesh::intersects(Position& r0, Position r1, int* ijk) const
+    }
+
+    // locations of fine meshes in y direction
+    for (int i=0; i<coarse_y.size(); i++) {
+      if (i==0) { 
+        width_y.push_back( (coarse_y.at(0)-lower_left_point[1])/shape_y.at(0) );
+        for (int j=0; j<shape_y.at(i); j++)  fine_y.push_back( lower_left_point[1]+width_y.back()*j );
+      } else {
+        width_y.push_back( (coarse_y.at(i)-coarse_y.at(i-1))/shape_y.at(i) );
+        for (int j=0; j<shape_y.at(i); j++)  fine_y.push_back( coarse_y.at(i-1)+width_y.back()*j );
+      }
+    }
+    fine_y.push_back(coarse_y.back());
+    
+    // Locations of the coarse meshes in z direction
+    if (check_for_node(node, "zmesh")) {
+      auto value = get_node_xarray<double>(node, "zmesh");
+      for (int i=0; i<value.size(); i++)  coarse_z.push_back(value.at(i));
+    }
+
+    // Number of fine meshes within corresponding coarse meshes in z direction
+    if (check_for_node(node, "zints")) {
+      auto value = get_node_xarray<int>(node, "zints");
+      for (int i=0; i<value.size(); i++)  shape_z.push_back(value.at(i));
+      if (coarse_z.size() != shape_z.size())      fatal_error("The number of zmesh and zints must be same.");
+    }
+
+    // locations of fine meshes in z direction
+    for (int i=0; i<coarse_z.size(); i++) {
+      if (i==0) { 
+        width_z.push_back( (coarse_z.at(0)-lower_left_point[2])/shape_z.at(0) );
+        for (int j=0; j<shape_z.at(i); j++)  fine_z.push_back( lower_left_point[2]+width_z.back()*j );
+      } else {
+        width_z.push_back( (coarse_z.at(i)-coarse_z.at(i-1))/shape_z.at(i) );
+        for (int j=0; j<shape_z.at(i); j++)  fine_z.push_back( coarse_z.at(i-1)+width_z.back()*j );
+      }
+    }
+    fine_z.push_back(coarse_z.back());
+
+    mesh_ = std::make_unique<RectilinearMesh>(fine_x, fine_y, fine_z);
+    
+    // Energy group
+    if (check_for_node(node, "energy")) {
+      xml_node weightwindow_energy = node.child("energy");
+      // energy group for neutron
+      if (check_for_node(weightwindow_energy,"neutron")) {
+        n_ww=true; // turn on the flag
+        auto value = get_node_xarray<double>(weightwindow_energy, "neutron");
+        n_energy_group.push_back(0);
+        for (int i=0; i<value.size(); i++)  n_energy_group.push_back(value.at(i));
+      }
+             
+      // energy group for photon
+      if (check_for_node(weightwindow_energy,"photon")) {
+        if (!settings::photon_transport) { fatal_error("Photon transport is not on but weight window for photon is used"); }  // check if photon transport is on
+        p_ww=true; // turn on the flag
+        auto value = get_node_xarray<double>(weightwindow_energy, "photon");
+        p_energy_group.push_back(0);
+        for (int i=0; i<value.size(); i++)  p_energy_group.push_back(value.at(i));
+      }   
+    } else { fatal_error("Must assign energy group for weight window"); }  
+
+    // read wwinp file
+    std::ifstream wwfile; 
```

Comment:
It seems a little strange to partially read the weight windows in the XML file and partially in from the wwinp ASCII file. It would probably be best to provide the weight window values in the XML file as well. Our `get_node_array` methods should make it pretty straightforward to read those in. It might make the `settings.xml` file quite large, but we can handle that later if needed.

For now, we can simply expect that the weight window values follow the same ordering specified in the wwinp file specification from MCNP.

Reply:
Yes, it will make the `settings.xml` file quite large, so I prefer to read the weight windows from a separate file to make `settings.xml` smaller and easier to read.

Reply:
While it's true that the `settings.xml` file might become very large, keeping all of our input in the same XML format really helps us maintain the code and produce more readable input. The Python API and the test suite both rely on this input format as well. It provides a more readable format which is also considerably easier to parse in comparison to a custom format like the wwinp files.

To alleviate the concern about the size of the `settings.xml` I think that eventually we'll want to place these variance reduction parameters into a separate XML file (either `vr.xml` or `variance_reduction.xml`), but I won't ask you to take that on here :smiley:

Reply:
Yes, it will be better that a separate XML file for the variance reduction is introduced.

Reply:
So, knowing that these values will eventually be placed in an input file dedicated to variance reduction, would you be willing to move the weight window arrays into the `settings.xml` for the purposes of this PR?

Reply:
Ok, I will finish this latter. :)